### PR TITLE
Fix order of aliases when reasoning up to aliases in unification

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** bug fix / enhancement

This was found when working on compiling pattern-matching with small inversion (which produces chains of let-ins). I could not easily find a way to produce a test for master though (this requires specific interleaved dependencies).

Details:
- first commit fix the order of aliases (most recent variables come first but lookups were done on the converse assumption) and also relax the evar restrictions by allowing all intermediate names of a chain of let-ins to be used while projecting variables in unification
- second commit is about documenting the code and restructuring it a bit

Complexity should not be changed (up to depending on the length of chains of `let xi:=xj in` but they are presumably short ??).

### CODE REVIEW

**Comment by herbelin** [comment]
Informing @ppedrot as author of the last iteration on the management of let-ins in unification.

File: pretyping/evarsolve.ml [hunk 1]
```diff
@@ -336,11 +336,17 @@ let eq_alias a b = match a, b with
 | VarAlias id1, VarAlias id2 -> Id.equal id1 id2
 | _ -> false
 
-type 'a aliasing = 'a option * alias list
+(* A chain of let-in ended either by a declared variable or a non-variable term *)
+(* e.g. [x:=1;y:=x;z:=y] binds [z] to [([y;x],Inr 1)] *)
+(* and. [a:T;x:=a;y:=x;z:=y] binds [z] to [([y;x],Inl a)] *)
+type 'a aliasing = alias list * (alias, 'a) union
```
Comment (ppedrot) [inline pretyping/evarsolve.ml:342]: I'd prefer a dedicated algebraic type here so as to make the intent clear.

Comment (ppedrot) [reply pretyping/evarsolve.ml:342]: @herbelin ping

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
Out of curiosity I am running a bench before merging: https://gitlab.com/coq/coq/-/jobs/1151148616

**Comment by ppedrot** [comment]
```
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                             │                         │                                       │                                       │                         │                 │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD  PDIFF │ NEW  OLD  PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-fingroup │   48.08    48.38  -0.62 │   133545463690    133586829133  -0.03 │   186786237005    186746204271   0.02 │  486436   482912   0.73 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-math-classes │  203.63   204.45  -0.40 │   565596435042    564397156070   0.21 │   762735200279    762714061275   0.00 │  499968   502136  -0.43 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-fiat-core │  108.40   108.58  -0.17 │   305473009591    304054096926   0.47 │   415689008175    414256028646   0.35 │  487208   486220   0.20 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                  coq-stdlib │  387.41   388.04  -0.16 │  1076601913564   1076806577269  -0.02 │  1396812044700   1396837102056  -0.00 │  571844   571832   0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│             coq-fiat-crypto │ 4830.72  4837.32  -0.14 │ 13442734097816  13419468996224   0.17 │ 23671901001676  23671883419170   0.00 │ 2229188  2236416  -0.32 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                   coq-flocq │  146.16   146.30  -0.10 │   405534032036    405702206089  -0.04 │   501589920554    501522932867   0.01 │  829980   827856   0.26 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│ coq-rewriter-perf-SuperFast │  761.09   761.69  -0.08 │  2119516918471   2122261771481  -0.13 │  3291999515573   3290680053877   0.04 │ 1106416  1104632   0.16 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-mathcomp-character │  153.11   153.19  -0.05 │   425277314823    425764362048  -0.11 │   598683590509    598457119255   0.04 │  710496   717724  -1.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-solvable │  171.00   171.08  -0.05 │   475238088693    475393183223  -0.03 │   676793696411    676866036618  -0.01 │  686708   690588  -0.56 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│          coq-mathcomp-field │  223.72   223.78  -0.03 │   622550417545    622512700446   0.01 │   950413523718    950251086115   0.02 │  661276   662076  -0.12 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-mathcomp-odd-order │ 1058.55  1058.44   0.01 │  2944411573992   2945510861766  -0.04 │  4688392806581   4686260627425   0.05 │  916612   911844   0.52 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-bignums │   55.29    55.28   0.02 │   152254517088    152408949453  -0.10 │   206420609619    206369559273   0.02 │  472684   472568   0.02 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-verdi-raft │ 1174.41  1174.14   0.02 │  3270833814616   3271650914810  -0.02 │  4626255971832   4625536675248   0.02 │  904436   903860   0.06 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-engine-bench-lite │  329.85   329.71   0.04 │   928679674227    928243597089   0.05 │  1686527045627   1690205312075  -0.22 │ 3546720  3546552   0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-unimath │ 3960.24  3957.56   0.07 │ 11012381067706  11005357768596   0.06 │ 21207570072221  21210586025692  -0.01 │ 1126484  1126548  -0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-rewriter │  694.50   693.76   0.11 │  1932141498117   1930696613970   0.07 │  2929069222594   2926935049811   0.07 │ 1104516  1106640  -0.19 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│             coq-lambda-rust │ 1258.26  1256.89   0.11 │  3498014724734   3493765216445   0.12 │  5030901977993   5028160246679   0.05 │ 1102528  1096216   0.58 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-perennial │ 5825.68  5818.15   0.13 │ 16232122253653  16215115133514   0.10 │ 24549340901636  24548249117661   0.00 │ 3110688  3098332   0.40 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│        coq-mathcomp-algebra │  139.40   139.16   0.17 │   386457142995    386341481115   0.03 │   503276492996    503262440551   0.00 │  555472   556844  -0.25 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-fiat-parsers │  627.31   626.14   0.19 │  1745235561036   1740652966857   0.26 │  2652769444611   2651928660482   0.03 │ 3283968  3283628   0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                    coq-hott │  315.17   314.57   0.19 │   844616189408    844117468391   0.06 │  1357333566337   1357158978767   0.01 │  592148   590096   0.35 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                    coq-corn │ 1465.78  1462.70   0.21 │  4063081608747   4054037853629   0.22 │  5866969087001   5866980355046  -0.00 │  806788   810280  -0.43 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-bedrock2 │  202.68   202.20   0.24 │   561595073024    560945162894   0.12 │   921996673717    922524338944  -0.06 │  778080   778884  -0.10 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-fourcolor │ 2526.56  2519.52   0.28 │  7039464620842   7022155747415   0.25 │ 12253282134904  12255040531221  -0.01 │  737888   735668   0.30 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│  coq-performance-tests-lite │ 1554.80  1550.43   0.28 │  4342223351284   4331889408663   0.24 │  7199166251194   7198969619895   0.00 │ 1701920  1696176   0.34 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                   coq-color │  492.42   490.99   0.29 │  1361432891563   1358865219144   0.19 │  1718257028448   1716991548958   0.07 │ 1070764  1069404   0.13 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                  coq-geocoq │ 1382.52  1377.63   0.35 │  3831582206553   3820682995110   0.29 │  5636953029055   5635352789307   0.03 │  996572   996644  -0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                    coq-core │  107.61   106.80   0.76 │   301881060469    301150063692   0.24 │   455815011658    455737733277   0.02 │  267732   267728   0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                   coq-verdi │   99.26    98.47   0.80 │   272988259401    272373787342   0.23 │   385336650867    384379317851   0.25 │  548656   548436   0.04 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-mathcomp-ssreflect │   53.19    52.72   0.89 │   145495456255    145256784290   0.16 │   185715786659    185700329750   0.01 │  524640   524172   0.09 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-coqutil │   55.24    54.69   1.01 │   152135441830    151330203835   0.53 │   198177132618    197667980152   0.26 │  524276   524268   0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-coquelicot │   69.27    68.55   1.05 │   188154218092    187148895453   0.54 │   239170681919    238267882153   0.38 │  558424   562320  -0.69 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-coqprime │   79.28    78.30   1.25 │   216421835053    216172954018   0.12 │   306226796632    305341494548   0.29 │  764616   759744   0.64 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                         coq │    4.27     4.19   1.91 │    11385295549     11390892573  -0.05 │    22530878159     22521674645   0.04 │  294384   294576  -0.07 │   0    0    nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘
```
Looks like noise.

**Pull request merged**